### DISCUSSION TITLE
How to ensure High Availability in JDBC_PING cluster for Database

### DISCUSSION THREAD

**Discussion by Nihal987** [keycloak/keycloak#22164] [Q&A] [answered]
If there are multiple keycloak instances in a cluster and the cluster was setup with JDBC_PING then isn't the database a single point of failure? If the database is down then all the users and configurations stored in the database won't be accessible.

Doesn't this defeat the purpose of a clustered setup in the first place?

**Comment by adnanvrbic** [comment]
How would Keycloak work at all if the database is down? Not everything is in the cache.

**Reply by Nihal987** [reply]
Yes you're right but that wasn't the point of my question.
A clusters offers High Availability, so if one keycloak instance goes down another instance can service those requests, but it's not highly available if the database goes down (it's just like a single instance of keycloak).

I wanted to know if there was a type of solution where each instance of keycloak could use it's own database somehow to solve this vulnerability.

I know that you can create a separate cluster for your database, with a load balancer or proxy but again that's a separate whole other cluster. I was wondering if this was the only available option to ensure high availability or if I was missing something.

**Comment by vmuzikar** [comment]
Seems closely related to our cross-dc effort.
CC @keycloak/store

**Reply by ahus1** [reply, marked as answer]
@Nihal987 - your are correct that you'll need to prepare a highly available (HA) database setup to avoid the database being a single-point-of-failure. 

How to set up such a HA database is up to you, and it will depend highly on the database you choose and the infrastructure you run it on.

We're starting to document some resilience patterns in [Keycloak Benchmark](https://www.keycloak.org/keycloak-benchmark/resilience-guide/latest/client-failover), but we haven't yet published anything on databases yet. It might take some more months for that, it will never be comprehensive. At the moment we're focusing on Keycloak and Infinispan only.

**Reply by Nihal987** [reply]
Ah I see, thank you. Well I look forward to seeing what resilience patterns are described in the future.

**Reply by ahus1** [reply]
Feel free to contribute to this repository which a setup :-) once you have a solution for your scenario. Currently the recipes described there feature Kubernetes/OpenShift and AWS.

**Reply by Nihal987** [reply]
I'm experimenting with PostgreSQL and MySQL, I'll definitely post once I come up with a solution!

**Reply by Nihal987** [reply]
Hi @ahus1 sorry just curious, keycloak doesn't already have a solution to connect to multiple databases does it? I realize I might have misinterpreted your previous comment.

**Reply by ahus1** [reply]
@Nihal987 - while Keycloak supports multiple database vendors, it supports to connect only to one database instance at a time. 

Exceptions for that are customers creating their own[ user store providers](https://www.keycloak.org/docs/latest/server_development/#_user-storage-spi) (which use a different JDBC connection from the standard one). 

For the sake of completeness: Sometimes people also use Keycloak's federation capabilities to have user data stored somewhere else.
